### ISSUE TITLE
Disable forwarding redirect for peer's HTTP communication

### ISSUE THREAD

**Issue by ahrtr** [etcd-io/etcd#16969]
### What would you like to be added?

It's possible that etcd server may run into SSRF (Server-side request forgery) when adding a new member. If users provide a malicious peer URL, the existing etcd members may be redirected to other unexpected internal URL when getting the new member's version,

https://github.com/etcd-io/etcd/blob/a9cf27b169b1fc89bc0189ffaea7d21c52856b24/server/etcdserver/cluster_util.go#L250

We should disable the forwarding redirect for the endpoint `/version`, as there is no need to support it.
```
client := &http.Client{
    Transport: rt,
    Timeout:   timeout,
    CheckRedirect: func(req *http.Request, via []*http.Request) error {
        return http.ErrUseLastResponse
    },
}
```

We should also evaluate all peer's HTTP communication, and disable the redirect as well if needed.
https://github.com/etcd-io/etcd/blob/d0114cf4c0ab8b491e7ef5c8bda8222983d05c93/server/etcdserver/api/etcdhttp/peer.go#L61-L75

### Why is this needed?

Improve security

**Comment by jmhbnz** [comment]
Discussed during sig etcd triage meeting. Assigned to @ivanvc to investigate remaining etcd peer http interactions.

**Comment by ivanvc** [comment]
Happy to help. Can you assign it to me? Thanks

**Comment by lavacat** [comment]
Another area to check out are endpoints from `Transport.Hander`
https://github.com/etcd-io/etcd/blob/904c0769e9aaf4b676bda8ff2a866e558a40270b/server/etcdserver/api/rafthttp/transport.go#L157C1-L166C12

**Comment by ivanvc** [comment]
If I understand correctly, we need to update the HTTP client from the following locations:
* Promote Member: https://github.com/etcd-io/etcd/blob/3b37afe/server/etcdserver/cluster_util.go#L292
* Lease
  * Renew: https://github.com/etcd-io/etcd/blob/3b37afe/server/lease/leasehttp/http.go#L153
  * TTL: https://github.com/etcd-io/etcd/blob/3b37afe/server/lease/leasehttp/http.go#L213
 * Downgrade Enabled: https://github.com/etcd-io/etcd/blob/3b37afe/server/etcdserver/cluster_util.go#L362-365
 * Hash KV: https://github.com/etcd-io/etcd/blob/3b37afe/server/etcdserver/corrupt.go#L470

As for the raft routes, I think they're safe as `http.RoundTripper` doesn't follow redirects, and the `checkPostResponse` from raft's utils (https://github.com/etcd-io/etcd/blob/3b37afe/server/etcdserver/api/rafthttp/util.go#L85-L117), doesn't handle redirects, so it shouldn't follow a redirection.

There's also a request in `cluster_util.go`, to get the cluster from remote peers (https://github.com/etcd-io/etcd/blob/3b37afe/server/etcdserver/cluster_util.go#L72-75). Should we disallow redirections there too?

**Comment by jmhbnz** [comment]
Thanks @ivanvc - I can't see how these additional http clients could be abused but I also don't see the harm in just disabling following redirects for them as to the best of my knowledge they should not be needed by etcd. 

My vote would go to raising a pull request to update all of these instances to not follow redirects. Perhaps leave it a few days to see if maintainers or other contributors have any different views on it.

**Comment by ivanvc** [comment]
@ahrtr, @lavacat, any thoughts on this? Else, I'll go ahead and implement as @jmhbnz suggested :)

**Comment by ivanvc** [comment]
Hey @jmhbnz, I saw you added the backport v3.5 label. Should I cherry-pick my commit and create a new PR targeting the `release-3.5` branch?

**Comment by jmhbnz** [comment]
Hey @ivanvc - Yes given this was a security concern my view is we should backport this to `release-3.5`.  

Note - the codebase structure can be quite different going back to `release-3.5` and particularly `release-3.4` so we generally avoid any automated cherry pick and just add a manual commit as normal to complete backports.

I will re-open until we get the backport merged, thanks for your help on this @ivanvc 🙏🏻

**Comment by ivanvc** [comment]
@jmhbnz I think we can close the ticket now unless we want to backport it to another version.

**Comment by jmhbnz** [comment]
Hey @ahrtr do you think we should be trying to backport an update for this to 3.4? It is security related so probably, but defer to you as release manager for the branch.

**Comment by ahrtr** [comment]
> Hey @ahrtr do you think we should be trying to backport an update for this to 3.4? It is security related so probably, but defer to you as release manager for the branch.

Yes, please backport the fix to 3.4 as well. Thanks.

**Comment by ivanvc** [comment]
> Yes, please backport the fix to 3.4 as well. Thanks.

Oops I didn't see this before. I will do it.

**Comment by ivanvc** [comment]
@jmhbnz, I included the change you did to `getVersion` in my commit: 838cd9aa00aed3bc06261b2a856cb6d51a3771d0.
